Is the hard requirement for _Facility name_ really necessary? Could we make it that if none is specified in the CSV file, and there's only one on the device, users get imported there?

`deb` asset installed on Ubuntu 16.04 gives the following error when trying to import users with the `kolibri manage importusers` command, successfully used in the past with the same CSV file.

```
osboxes@osboxes:~$ cd Desktop/
osboxes@osboxes:~/Desktop$ sudo dpkg -i kolibri-buildkite-build-5953-12101-kolibri_0.13.0.dev0+git.20190925180238-0ubuntu1_all.deb 
[sudo] password for osboxes: 
Selecting previously unselected package kolibri.
(Reading database ... 334845 files and directories currently installed.)
Preparing to unpack kolibri-buildkite-build-5953-12101-kolibri_0.13.0.dev0+git.20190925180238-0ubuntu1_all.deb ...
Unpacking kolibri (0.13.0.dev0+git.20190925180238-0ubuntu1) ...
Setting up kolibri (0.13.0.dev0+git.20190925180238-0ubuntu1) ...
Synchronizing state of kolibri.service with SysV init with /lib/systemd/systemd-sysv-install...
Executing /lib/systemd/systemd-sysv-install enable kolibri
Processing triggers for systemd (229-4ubuntu21.22) ...
Processing triggers for ureadahead (0.100.0-19.1) ...
osboxes@osboxes:~/Desktop$ kolibri manage importusers many-users.csv 
INFO     Option RUN_MODE in section [Deployment] being overridden by environment variable KOLIBRI_RUN_MODE
INFO     Running Kolibri with the following settings: kolibri.deployment.default.settings.base
INFO     Invoking command importusers many-users.csv
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/kolibri/core/auth/csv_utils.py", line 24, in infer_facility
    facility = Facility.objects.get(pk=facility_id)
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/db/models/query.py", line 374, in get
    num = len(clone)
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/db/models/query.py", line 232, in __len__
    self._fetch_all()
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/db/models/query.py", line 1121, in _fetch_all
    self._result_cache = list(self._iterable_class(self))
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/db/models/query.py", line 53, in __iter__
    results = compiler.execute_sql(chunked_fetch=self.chunked_fetch)
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/db/models/sql/compiler.py", line 876, in execute_sql
    sql, params = self.as_sql()
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/db/models/sql/compiler.py", line 441, in as_sql
    where, w_params = self.compile(self.where) if self.where is not None else ("", [])
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/db/models/sql/compiler.py", line 373, in compile
    sql, params = node.as_sql(self, self.connection)
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/db/models/sql/where.py", line 79, in as_sql
    sql, params = compiler.compile(child)
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/db/models/sql/compiler.py", line 373, in compile
    sql, params = node.as_sql(self, self.connection)
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/db/models/lookups.py", line 170, in as_sql
    rhs_sql, rhs_params = self.process_rhs(compiler, connection)
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/db/models/lookups.py", line 103, in process_rhs
    return self.get_db_prep_lookup(value, connection)
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/db/models/lookups.py", line 196, in get_db_prep_lookup
    [get_db_prep_value(value, connection, prepared=True)]
  File "/usr/lib/python3/dist-packages/kolibri/dist/morango/models/fields/uuids.py", line 38, in get_db_prep_value
    value = uuid.UUID(value)
  File "/usr/lib/python3.5/uuid.py", line 140, in __init__
    raise ValueError('badly formed hexadecimal UUID string')
ValueError: badly formed hexadecimal UUID string

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/kolibri/core/auth/csv_utils.py", line 27, in infer_facility
    facility = Facility.objects.get(name=facility_id)
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/db/models/query.py", line 380, in get
    self.model._meta.object_name
kolibri.core.auth.models.DoesNotExist: Facility matching query does not exist.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/bin/kolibri", line 11, in <module>
    load_entry_point('kolibri==0.13.0.dev0+git.20190925180238', 'console_scripts', 'kolibri')()
  File "/usr/lib/python3/dist-packages/kolibri/dist/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/kolibri/dist/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/usr/lib/python3/dist-packages/kolibri/dist/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/lib/python3/dist-packages/kolibri/utils/cli.py", line 235, in invoke
    return super(KolibriDjangoCommand, self).invoke(ctx)
  File "/usr/lib/python3/dist-packages/kolibri/dist/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/lib/python3/dist-packages/kolibri/dist/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/kolibri/dist/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/usr/lib/python3/dist-packages/kolibri/utils/cli.py", line 633, in manage
    execute_from_command_line(["kolibri manage"] + ctx.args)
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/core/management/__init__.py", line 364, in execute_from_command_line
    utility.execute()
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/core/management/__init__.py", line 356, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/core/management/base.py", line 283, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/usr/lib/python3/dist-packages/kolibri/dist/django/core/management/base.py", line 330, in execute
    output = self.handle(*args, **options)
  File "/usr/lib/python3/dist-packages/kolibri/core/auth/management/commands/importusers.py", line 210, in handle
    create_user(map_input(row), default_facility=default_facility)
  File "/usr/lib/python3/dist-packages/kolibri/core/auth/management/commands/importusers.py", line 89, in create_user
    facility = infer_facility(user.get("facility", None), facility=default_facility)
  File "/usr/lib/python3/dist-packages/kolibri/core/auth/csv_utils.py", line 31, in infer_facility
    facility=facility_id
ValueError: Facility matching identifier  was not found
osboxes@osboxes:~/Desktop$ 
```


The structure of the CSV file is as follows:
```
full_name,username,password,facility,class
Marc G.,marcg,marcg,,Primera
Peter B.,peterb,peterb,,Primera
Neela R.,neelar,neelar,,Primera
...
Archie M.2,archiem2,archiem2,,Segunda
Archie M.3,archiem3,archiem3,,Segunda
Archie M.4,archiem4,archiem4,,Segunda
```

[many-users.zip](https://github.com/learningequality/kolibri/files/3654826/many-users.zip)

--- a/kolibri/core/auth/csv_utils.py
+++ b/kolibri/core/auth/csv_utils.py
@@ -8,8 +8,16 @@ import sys
 from collections import OrderedDict
 from functools import partial
 
+from django.db.models import IntegerField
+from django.db.models import OuterRef
+from django.db.models import Prefetch
+from django.db.models import Subquery
+
+from kolibri.core.auth.constants.collection_kinds import CLASSROOM
 from kolibri.core.auth.constants.demographics import choices
 from kolibri.core.auth.constants.demographics import DEMO_FIELDS
+from kolibri.core.auth.models import Classroom
+from kolibri.core.auth.models import Collection
 from kolibri.core.auth.models import Facility
 from kolibri.core.auth.models import FacilityUser
 
@@ -18,7 +26,7 @@ logger = logging.getLogger(__name__)
 
 
 def infer_facility(facility_id, facility=None):
-    if facility_id is not None:
+    if facility_id:
         try:
             # Try lookup by id first, then name
             facility = Facility.objects.get(pk=facility_id)
@@ -55,9 +63,24 @@ def transform_choices(field, obj):
     return choices_dict.get(obj[field], obj[field])
 
 
-mappings = {
+MULTIPLE_CLASSROOMS_TEXT = "User is enrolled in multiple classrooms"
+
+
+def replace_multiple_classrooms(field, obj):
+    if "classroom_count" in obj and obj["classroom_count"] > 1:
+        return MULTIPLE_CLASSROOMS_TEXT
+    return obj[field]
+
+
+output_mappings = {
     "gender": partial(transform_choices, "gender"),
     "birth_year": partial(transform_choices, "birth_year"),
+    "memberships__collection__id": partial(
+        replace_multiple_classrooms, "memberships__collection__id"
+    ),
+    "memberships__collection__name": partial(
+        replace_multiple_classrooms, "memberships__collection__name"
+    ),
 }
 
 labels = OrderedDict(
@@ -67,13 +90,24 @@ labels = OrderedDict(
         ("memberships__collection__name", "Class name"),
         ("memberships__collection__id", "Class id"),
         ("full_name", "Full name"),
+        ("username", "Username"),
+        ("password", "Password"),
         ("gender", "Gender"),
         ("birth_year", "Birth year"),
         ("id_number", "ID number"),
-        ("username", "Username"),
-        ("password", "Password"),
     )
 )
+
+
+def map_output(obj):
+    mapped_obj = {}
+    for header, label in labels.items():
+        if header in output_mappings and header in obj:
+            mapped_obj[label] = output_mappings[header](obj)
+        elif header in obj:
+            mapped_obj[label] = obj[header]
+    return mapped_obj
+
 
 input_fields = (
     "full_name",
@@ -100,8 +134,14 @@ def transform_inputs(field, obj):
     return input_choices.get(obj[field], obj[field])
 
 
+def map_class(obj):
+    value = get_field(["class", "Class id", "Class name"], obj)
+    if value != MULTIPLE_CLASSROOMS_TEXT:
+        return value
+
+
 input_mappings = {
-    "class": partial(get_field, ["class", "Class id", "Class name"]),
+    "class": map_class,
     "facility": partial(get_field, ["facility", "Facility id", "Facility name"]),
 }
 
@@ -119,27 +159,24 @@ def map_input(obj):
     return mapped_obj
 
 
-def map_object(obj):
-    mapped_obj = {}
-    for header, label in labels.items():
-        if header in mappings and header in obj:
-            mapped_obj[label] = mappings[header](obj)
-        elif header in obj:
-            mapped_obj[label] = obj[header]
-    return mapped_obj
-
-
 db_columns = (
     "facility__name",
     "facility__id",
     "memberships__collection__name",
     "memberships__collection__id",
+    "classroom_count",
     "full_name",
     "username",
     "gender",
     "birth_year",
     "id_number",
 )
+
+
+class SQCount(Subquery):
+    # Include ALIAS at the end to support Postgres
+    template = "(SELECT COUNT(%(field)s) FROM (%(subquery)s) AS %(field)s__sum)"
+    output_field = IntegerField()
 
 
 def csv_file_generator(facility, filepath, overwrite=True, demographic=False):
@@ -166,8 +203,24 @@ def csv_file_generator(facility, filepath, overwrite=True, demographic=False):
         writer = csv.DictWriter(f, header_labels)
         logger.info("Creating csv file {filename}".format(filename=filepath))
         writer.writeheader()
-        for item in queryset.select_related(
-            "facility", "memberships__collection"
-        ).values(*columns):
-            writer.writerow(map_object(item))
-            yield
+        usernames = set()
+        for item in (
+            queryset.select_related("facility")
+            .annotate(
+                classroom_count=SQCount(
+                    Classroom.objects.filter(membership__user=OuterRef("id")),
+                    field="id",
+                )
+            )
+            .prefetch_related(
+                Prefetch(
+                    "memberships__collection",
+                    queryset=Collection.objects.filter(kind=CLASSROOM),
+                )
+            )
+            .values(*columns)
+        ):
+            if item["username"] not in usernames:
+                writer.writerow(map_output(item))
+                usernames.add(item["username"])
+                yield

--- a/kolibri/core/auth/csv_utils.py
+++ b/kolibri/core/auth/csv_utils.py
@@ -10,14 +10,13 @@ from functools import partial
 
 from django.db.models import IntegerField
 from django.db.models import OuterRef
-from django.db.models import Prefetch
+from django.db.models import Q
 from django.db.models import Subquery
 
 from kolibri.core.auth.constants.collection_kinds import CLASSROOM
 from kolibri.core.auth.constants.demographics import choices
 from kolibri.core.auth.constants.demographics import DEMO_FIELDS
 from kolibri.core.auth.models import Classroom
-from kolibri.core.auth.models import Collection
 from kolibri.core.auth.models import Facility
 from kolibri.core.auth.models import FacilityUser
 
@@ -212,13 +211,12 @@ def csv_file_generator(facility, filepath, overwrite=True, demographic=False):
                     field="id",
                 )
             )
-            .prefetch_related(
-                Prefetch(
-                    "memberships__collection",
-                    queryset=Collection.objects.filter(kind=CLASSROOM),
-                )
-            )
+            .prefetch_related("memberships__collection")
             .values(*columns)
+            .filter(
+                Q(memberships__collection__kind=CLASSROOM)
+                | Q(memberships__collection__isnull=True)
+            )
         ):
             if item["username"] not in usernames:
                 writer.writerow(map_output(item))

--- a/kolibri/core/auth/csv_utils.py
+++ b/kolibri/core/auth/csv_utils.py
@@ -212,11 +212,11 @@ def csv_file_generator(facility, filepath, overwrite=True, demographic=False):
                 )
             )
             .prefetch_related("memberships__collection")
-            .values(*columns)
             .filter(
                 Q(memberships__collection__kind=CLASSROOM)
                 | Q(memberships__collection__isnull=True)
             )
+            .values(*columns)
         ):
             if item["username"] not in usernames:
                 writer.writerow(map_output(item))

--- a/kolibri/core/auth/management/commands/importusers.py
+++ b/kolibri/core/auth/management/commands/importusers.py
@@ -89,14 +89,18 @@ def create_user(user, default_facility=None):
     facility = infer_facility(user.get("facility", None), facility=default_facility)
     classroom = infer_and_create_class(user.get("class", None), facility)
     username = user["username"]
+    password = user.get("password", "")
     try:
         user_obj = FacilityUser.objects.get(username=username, facility=facility)
+        if password:
+            user_obj.set_password(password)
+            user_obj.save()
         update_user_demographics(user, user_obj)
 
         if classroom and not user_obj.is_member_of(classroom):
             classroom.add_member(user_obj)
             logger.info(
-                'Exiting user "{username}" was added to a classroom "{classroom}"'.format(
+                'Existing user "{username}" was added to a classroom "{classroom}"'.format(
                     username=username, classroom=classroom.name
                 )
             )
@@ -202,7 +206,7 @@ class Command(BaseCommand):
             if has_header:
                 reader = csv.DictReader(f, strict=True)
             else:
-                reader = csv.reader(f, strict=True)
+                reader = csv.DictReader(f, fieldnames=input_fields, strict=True)
             with transaction.atomic():
                 total = 0
                 for row in reader:

--- a/kolibri/core/auth/test/test_user_export.py
+++ b/kolibri/core/auth/test/test_user_export.py
@@ -17,7 +17,9 @@ from kolibri.core.auth.constants.demographics import MALE
 from kolibri.core.auth.constants.demographics import NOT_SPECIFIED
 from kolibri.core.auth.csv_utils import labels
 from kolibri.core.auth.csv_utils import transform_choices
+from kolibri.core.auth.models import Classroom
 from kolibri.core.auth.models import FacilityUser
+from kolibri.core.auth.models import LearnerGroup
 
 
 users = [
@@ -43,7 +45,7 @@ users = [
 ]
 
 
-class ContentSummaryLogCSVExportTestCase(TestCase):
+class UserCSVExportTestCase(TestCase):
     def test_csv_export_with_demographics(self):
         facility, superuser = setup_device()
         for user in users:
@@ -98,3 +100,56 @@ class ContentSummaryLogCSVExportTestCase(TestCase):
         for demo_field in DEMO_FIELDS:
             label = labels[demo_field]
             self.assertFalse(label in results[0])
+
+    def test_csv_export_user_in_multiple_classes(self):
+        facility, superuser = setup_device()
+        classroom1 = Classroom.objects.create(name="testclass1", parent=facility)
+        classroom2 = Classroom.objects.create(name="testclass2", parent=facility)
+        user = users[0]
+        user_obj = FacilityUser.objects.create(facility=facility, **user)
+        classroom1.add_member(user_obj)
+        classroom2.add_member(user_obj)
+        _, filepath = tempfile.mkstemp(suffix=".csv")
+        call_command(
+            "exportusers", output_file=filepath, overwrite=True, demographic=True
+        )
+        if sys.version_info[0] < 3:
+            csv_file = open(filepath, "rb")
+        else:
+            csv_file = open(filepath, "r", newline="")
+        with csv_file as f:
+            results = list(row for row in csv.DictReader(f))
+
+        self.assertEqual(len(results), 2)
+
+    def test_csv_export_user_in_one_class_one_group(self):
+        facility, superuser = setup_device()
+        classroom = Classroom.objects.create(name="testclass", parent=facility)
+        group = LearnerGroup.objects.create(name="testgroup", parent=classroom)
+        user = users[0]
+        user_obj = FacilityUser.objects.create(facility=facility, **user)
+        classroom.add_member(user_obj)
+        group.add_member(user_obj)
+        _, filepath = tempfile.mkstemp(suffix=".csv")
+        call_command(
+            "exportusers", output_file=filepath, overwrite=True, demographic=True
+        )
+        if sys.version_info[0] < 3:
+            csv_file = open(filepath, "rb")
+        else:
+            csv_file = open(filepath, "r", newline="")
+        with csv_file as f:
+            results = list(row for row in csv.DictReader(f))
+
+        self.assertEqual(len(results), 2)
+
+        for row in results:
+            user = filter(lambda x: x["username"] == row[labels["username"]], users)
+            if user:
+                user = user[0]
+                self.assertEqual(
+                    row[labels["memberships__collection__name"]], classroom.name
+                )
+                self.assertEqual(
+                    row[labels["memberships__collection__id"]], classroom.id
+                )

--- a/kolibri/core/auth/test/test_user_export.py
+++ b/kolibri/core/auth/test/test_user_export.py
@@ -63,9 +63,10 @@ class UserCSVExportTestCase(TestCase):
             results = list(row for row in csv.DictReader(f))
 
         for row in results:
-            user = filter(lambda x: x["username"] == row[labels["username"]], users)
+            user = next(
+                filter(lambda x: x["username"] == row[labels["username"]], users), None
+            )
             if user:
-                user = user[0]
                 self.assertEqual(row[labels["birth_year"]], user["birth_year"])
                 self.assertEqual(
                     row[labels["gender"]], transform_choices("gender", user)
@@ -144,9 +145,10 @@ class UserCSVExportTestCase(TestCase):
         self.assertEqual(len(results), 2)
 
         for row in results:
-            user = filter(lambda x: x["username"] == row[labels["username"]], users)
+            user = next(
+                filter(lambda x: x["username"] == row[labels["username"]], users), None
+            )
             if user:
-                user = user[0]
                 self.assertEqual(
                     row[labels["memberships__collection__name"]], classroom.name
                 )

--- a/kolibri/core/auth/test/test_user_import.py
+++ b/kolibri/core/auth/test/test_user_import.py
@@ -43,6 +43,10 @@ class UserImportTestCase(TestCase):
         default = {}
         self.assertEqual(infer_facility(None, default), default)
 
+    def test_infer_facility_empty_string(self):
+        default = {}
+        self.assertEqual(infer_facility("", default), default)
+
     def test_infer_facility_by_id(self):
         default = {}
         self.assertEqual(infer_facility(self.facility.id, default), self.facility)

